pkg/vmconfig/virtualcontroller/virtualcontroller_reconciler.go

```diff
@@ -179,6 +173,24 @@ func (r reconciler) Reconcile(
 			"DeviceChange", configSpec.DeviceChange[oldDeviceChangeCount:])
 	}
 
+	// Only allow add/delete ops when VM is poweredOn, not edit.
+	if slices.ContainsFunc(
+		configSpec.DeviceChange[oldDeviceChangeCount:],
+		func(bdc vimtypes.BaseVirtualDeviceConfigSpec) bool {
+			dc := bdc.GetVirtualDeviceConfigSpec()
+			return dc != nil &&
+				dc.Operation == vimtypes.VirtualDeviceConfigSpecOperationEdit
+		},
+	) {
+		if moVM.Runtime.PowerState != vimtypes.VirtualMachinePowerStatePoweredOff {
+			pkglog.FromContextOrDefault(ctx).Info("hot-update virtual controllers when VM is not powered off is not allowed, skip updating")
+			// Clear any new updates from DeviceChange.
+			configSpec.DeviceChange = configSpec.DeviceChange[:oldDeviceChangeCount]
```

Comment:
Right now I cleared all updates, including `add`/`remove`. Let me know if you think we should leave those and only clears `Edit`